hydro_deploy/core/src/ssh.rs

```diff
@@ -75,81 +80,85 @@ impl LaunchedBinary for LaunchedSshBinary {
         }
     }
 
-    async fn wait(&mut self) -> Option<i32> {
+    async fn wait(&mut self) -> Result<i32> {
         self.channel.wait_eof().await.unwrap();
-        let ret = self.exit_code();
+        let exit_code = self.channel.exit_status()?;
         self.channel.wait_close().await.unwrap();
-        ret
+
+        Ok(exit_code)
+    }
+
+    async fn stop(&mut self) -> Result<()> {
+        if !self.channel.eof() {
+            self.channel.write_all(b"\x03").await?; // `^C`
+            self.channel.send_eof().await?;
+            self.channel.wait_eof().await?;
+            // `exit_status()`
+            self.channel.wait_close().await?;
+        }
+
+        // Run perf post-processing and download perf output.
+        if let Some(perf) = &self.perf {
+            let local_file = perf.output_file.to_str().unwrap();
+            let mut local_file = tokio::fs::File::create(local_file).await?;
+            {
+                let mut script_channel = async_retry(
+                    || async { Ok(self.session.as_ref().unwrap().channel_session().await?) },
+                    10,
+                    Duration::from_secs(1),
+                )
+                .await?;
+                let mut stderr_lines = BufReader::new(script_channel.stderr()).lines();
+                let stderr_task = tokio::task::spawn(async move {
+                    while let Some(Ok(s)) = stderr_lines.next().await {
+                        ProgressTracker::println(&format!("[perf stderr] {s}"));
+                    }
+                });
+                let mut stdout = script_channel.stream(0).compat();
+                let stdout_task = tokio::task::spawn(async move {
+                    // Download output via stdout.
+                    // Writing to a file and downloading via `sftp` may be faster as it can download non-sequentially.
+                    tokio::io::copy(&mut stdout, &mut local_file).await
```

Comment:
Oh nice, this also sets us up for in-memory processing by just `read_all`ing from `stdout`.